static/js/theme-map/Util/helpers.js

```diff
@@ -1,30 +1,37 @@
 import { GoogleMaps } from '../Maps/Providers/Google.js';
 import { MapboxMaps } from '../Maps/Providers/Mapbox.js';
+import { parseLocale } from '../../utils.js';
 
 /**
- * Gets the language locale according to specific fallback logic
- * 1. The user-specified locale to the component
- * 2. If invalid, try using only the first two characters
- * 3. If still invalid, providers fallback to en
+ * Gets the language locale according to user-specified locale to the component.
+ * If language code is not 2 character, default to en. If map provider doesn't
+ * support the specific language locale (with modifier and region), then default
+ * to the 2 character language code.
  *
  * @param {string} localeStr The user-defined locale string
  * @param {string} mapProvider name of the current map provider
  * @return {string} The language locale for the map
  */
 const getLanguageForProvider = (localeStr, mapProvider) => {
-  if (localeStr.length == 2) {
-    return localeStr;
-  } 
-
-  if (localeStr.length > 2) {
-    const provider = getMapProvider(mapProvider);
-    const formattedLocaleStr = localeStr.replace('_', '-');
-    if (provider.getSupportedLocales().includes(formattedLocaleStr)) {
-      return formattedLocaleStr;
+  const { language, modifier, region } = parseLocale(localeStr);
+  if (language.length !== 2) {
+    return 'en';
+  }
+  const provider = getMapProvider(mapProvider);
+  let formattedLocaleStr = modifier ? `${language}-${modifier}` : language;
+  if (provider === GoogleMaps) {
+    if (region) {
+      formattedLocaleStr = `${language}-${region}`;
+    } else if (modifier === 'Hant') {
+      formattedLocaleStr = `${language}-TW`;
+    } else if (modifier === 'Hans') {
+      formattedLocaleStr = `${language}-CN`;
     }
-    return localeStr.substring(0, 2);
   }
-  return 'en';
+  if (provider.getSupportedLocales().includes(formattedLocaleStr)) {
+    return formattedLocaleStr;
+  }
+  return language;
```

Comment:
just double checking we want to return just `zh` if say `zh-hans-cn` is not supported

Reply:
are you referring to the case where `zh-hans` is supported but not `zh-hans-cn`, we should return `zh`?

Reply:
oh I see formattedLocaleStr never is equal to something like `zh-hans-cn`, would it be possible to move the GoogleMaps specific logic here into the provider itself? Maybe a `transformLocaleForProvider` or something

Reply:
In the SDK, `zh-hans-cn` would return `zh-CN` for google maps. I think we should have the same behavior here

Reply:
yeah I misread the code @cea2aj I assumed formattedLocale would just be the locale canonicalized jambo style